californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -225,131 +237,174 @@ public Exchange receiveResponse(final Response response) {
+				final Exchange prev = exchangeStore.find(idByMID);
 				if (prev != null) {
-					try {
-						if (endpointContextMatcher.isResponseRelatedToRequest(prev.getEndpointContext(),
-								response.getSourceContext())) {
-							LOGGER.trace("received response for already completed {}: {}", prev, response);
-							response.setDuplicate(true);
-							return prev;
+					prev.execute(new Runnable() {
+
+						@Override
+						public void run() {
+
+							try {
+								if (endpointContextMatcher.isResponseRelatedToRequest(prev.getEndpointContext(),
+										response.getSourceContext())) {
+									LOGGER.trace("received response for already completed {}: {}", prev, response);
+									response.setDuplicate(true);
+									receiver.receiveResponse(prev, response);
+									return;
+								}
+							} catch (Exception ex) {
+								LOGGER.error("error receiving response {} for {}", response, prev, ex);
+							}
+							reject(response, receiver);
 						}
-					} catch (Exception ex) {
-						LOGGER.error("error receiving response {} for {}", response, prev, ex);
-					}
+					});
+				} else {
+					reject(response, receiver);
 				}
 			} else {
 				LOGGER.trace("discarding unmatchable piggy-backed response from [{}]: {}", response.getSourceContext(),
 						response);
 			}
-			// ignore response
-			return null;
+			return;
 		}
 
-		EndpointContext context = exchange.getEndpointContext();
-		Request currentRequest = exchange.getCurrentRequest();
-		int requestMid = currentRequest.getMID();
-		if (context == null) {
-			LOGGER.debug("ignoring response {}, request pending to sent!", response);
-			return null;
-		}
+		final Exchange exchange = tempExchange;
+		exchange.execute(new Runnable() {
 
-		try {
-			if (endpointContextMatcher.isResponseRelatedToRequest(context, response.getSourceContext())) {
-				if (response.getType() == Type.ACK && requestMid != response.getMID()) {
-					// The token matches but not the MID.
-					LOGGER.warn("possible MID reuse before lifetime end for token {}, expected MID {} but received {}",
-							response.getTokenString(), requestMid, response.getMID());
-					// when nested blockwise request/responses occurs (e.g.
-					// caused by retransmission), a old response may stop the
-					// retransmission of the current blockwise request. This
-					// seems to be a side effect of reusing the token. If the
-					// response to this current request is lost, the blockwise
-					// transfer times out, because the retransmission is stopped
-					// too early. Therefore don't return a exchange when the MID
-					// doesn't match.
-					// See issue #275
-					return null;
+			@Override
+			public void run() {
+				boolean checkResponseToken = !exchange.isNotification() || exchange.getRequest() != exchange.getCurrentRequest();
+				if (checkResponseToken && exchangeStore.get(idByToken) != exchange) {
+					LOGGER.debug("ignoring response {}, exchange not longer matching!", response);
+					return;
```

Comment:
not sure to understand this test ?

Reply:
the one part of the check is simple:

`exchangeStore.get(idByToken) != exchange`

Is this exchange still be the on returned for this token? 
Or has this changed in the meantime?

The other part of the check is, when this rule should be applied:

`!exchange.isNotification() || exchange.getRequest() != exchange.getCurrentRequest()`

Check it only, if it is no notification exchange (the check would fail, becaus it's in the observation-store), or if it is a blockwise exchange of the left blocks, which then again, is put into the `exchangeStore`.

I had also a version, which checks the notifies in the observation store. But I'm not sure, if that pays off.

I also think, that I need to extend the benchmark client with more statistic, to ensure, that this doesn't lower the quality (e.g. of blockwise notifies).